atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -324,24 +336,92 @@ private synchronized void refreshPool() {
 
-        serversToAdd.forEach(server -> cassandra.returnOrCreatePool(server, absentHostTracker.returnPool(server)));
         absentServers.forEach(cassandraServer -> {
             CassandraClientPoolingContainer container = cassandra.removePool(cassandraServer);
             absentHostTracker.trackAbsentCassandraServer(cassandraServer, container);
         });
 
         Set<CassandraServer> serversToShutdown = absentHostTracker.incrementAbsenceAndRemove();
 
-        if (!(serversToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
-            sanityCheckRingConsistency();
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(getCurrentPools(), serversToAdd);
+
+        if (!(validatedServersToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
             cassandra.refreshTokenRangesAndGetServers();
         }
 
-        logRefreshedHosts(serversToAdd, serversToShutdown, absentServers);
+        Preconditions.checkState(
+                !getCurrentPools().isEmpty() || serversToAdd.isEmpty(),
+                "No servers were successfully added to the pool. This means we could not come to a consensus on"
+                    + " cluster topology, and the client cannot start as there are no valid hosts. This state should"
+                    + " be transient (<5 minutes), and if it is not, indicates that the user may have accidentally"
+                    + " configured AltasDB to use two separate Cassandra clusters (i.e., user-led split brain).",
+                SafeArg.of("serversToAdd", serversToAdd));
+
+        logRefreshedHosts(validatedServersToAdd, serversToShutdown, absentServers);
+    }
+
+    /**
+     * Validates new servers to add to the cassandra client container pool,
+     * by checking them with the @see {@link com.palantir.atlasdb.keyvalue.cassandra.CassandraTopologyValidator}.
+     * If any servers come back and are not in consensus this is OK, we will simply add them to the absent host
+     * tracker, as we most likely will retry this host in subsequent calls.
+     *
+     * @return The set of cassandra servers which have valid topologies and have been added to the pool.
+     */
+    @VisibleForTesting
+    Set<CassandraServer> validateNewHostsTopologiesAndMaybeAddToPool(
+            Map<CassandraServer, CassandraClientPoolingContainer> currentContainers,
+            Set<CassandraServer> serversToAdd) {
+        if (serversToAdd.isEmpty()) {
+            return Set.of();
+        }
+
+        Map<CassandraServer, CassandraClientPoolingContainer> serversToAddContainers =
+                getContainerForNewServers(serversToAdd);
+
+        Preconditions.checkArgument(
+                Sets.intersection(currentContainers.keySet(), serversToAddContainers.keySet())
+                        .isEmpty(),
+                "The current pool of servers should not have any server(s) that are being added.",
```

Comment:
minor nit: this is declaring what should be the correct argument, as opposed to what the state is when this exception is thrown (i.e., closer to something like Tried adding a server that already exists. This indicates a bug in.... )